DOCSP-51344: Logging (#115)
mongodb/docs-kotlin-sync: source/logging-and-monitoring/logging.txt

   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to set up and configure a logger in the {+driver-short+}.

This guide shows how to record events in the driver. If you want to learn how to
use information about the activity of the driver in code, see the :ref:`kotlin-sync-monitoring`
guide.

Set Up a Logger
---------------

> The {+driver-short+} uses SLF4J to allow you to specify your logging framework of choice.

Reviewer: ```suggestion
The {+driver-short+} uses Simple Logging Facade For Java (SLF4J) to allow you to specify your logging framework of choice.
```
I think it is good to list the full name once

         12:14:55.861 [main] DEBUG org.mongodb.driver.protocol.command - Command "find" started on database <database> using a connection with driver-generated ID 3 and server-generated ID 3 to <MongoDB hostname>. The request ID is 5. Command: {"find": "<collection>", "filter": {}, "limit": 1, "singleBatch": true, "$db": "<database>", "lsid": {"id": {"$binary": {"base64": "<_id>", "subType": "04"}}}, "$readPreference": {"mode": "primaryPreferred"}}
         12:14:55.864 [main] DEBUG org.mongodb.driver.protocol.command - Command "find" succeeded in 4.34 ms using a connection with driver-generated ID 3 and server-generated ID 3 to <MongoDB hostname. The request ID is 5. Command reply: {"cursor": {"id": 0, "ns": "<database>.<collection>", "firstBatch": []}, "ok": 1.0, "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1673778535, "i": 1}}, "signature": {"hash": {"$binary": {"base64": "<_id>", "subType": "00"}}, "keyId": 0}}, "operationTime": {"$timestamp": {"t": 1673778535, "i": 1}}}

      .. note:: Default Log Level

         The default log level of Logback is DEBUG. To learn how to change your
         Logback logger's log level, see the

>          :ref:`Configure Log Level <kotlin-sync-configure-log-level>` section of this page.

Reviewer: suggestion: this happens to me, but once the page loads from the link it does not scroll to the example but rather stays in the same section. to prevent confusion, maybe use copy link to highlight even though it clogs the rst.

Author: Hmm, I'm not seeing this bug. Section links are known to be a little finicky, but this is a DOP-level issue. I'd prefer to be consistent with our other section links.

Reviewer: Sounds good, it could be my browser or wifi doing this, as it happens on my PRs too

      Log4j2 with your project's dependency management tool.

      .. tabs::

         .. tab:: Maven
            :tabid: maven

            Add the following dependency to your ``pom.xml`` file.

            .. include:: /includes/logging-and-monitoring/log4j2-maven.rst

         .. tab:: Gradle
            :tabid: gradle

            Add the following dependency to your ``build.gradle`` file.

            .. include:: /includes/logging-and-monitoring/log4j2-gradle.rst

      Once you have included the preceding dependency, log an error by using the
      following code:

      .. code-block:: kotlin

         import org.slf4j.Logger;
         import org.slf4j.LoggerFactory;

         ...

         val logger = LoggerFactory.getLogger("MyApp");
         logger.error("Logging an Error");

      The preceding code will ouput a message that resembles the following:

      .. code-block:: none
         :copyable: false

         12:35:00.438 [main] ERROR <my package path> - Logging an Error

      .. note:: Default Log Level

         The default log level of Log4J2 is ERROR. This means that running

>          standard operations in the MongoDB Java driver will not produce output

Reviewer: ```suggestion
         standard operations in the {+driver-long+} will not produce output
```

               </Console>
            </Appenders>
            <Loggers>
               <Root level="INFO">
                     <AppenderRef ref="Console"/>
               </Root>
            </Loggers>
         </Configuration>

      To test that your logger configuration was successful, run the following
      code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none
         :copyable: false

         ...
         10:14:57.633 [cluster-ClusterId{value=<your cluster id>, description='null'}-<your connection uri>] INFO  org.mongodb.driver.cluster - Discovered replica set primary <your connection uri>
         10:14:57.790 [main] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:7, serverValue:<your server value>}] to <your connection uri>

Logger Names
------------

Your logger uses logger names to help organize different logging events. Logger names
are strings that form a hierarchy. A logger is an ancestor of another logger if its
name followed by a ``"."`` is a prefix of the other logger's name. For example,

> ``"grandparent"`` is an ancestor of ``"grandparent.parent"`` which is an ancestor of

Reviewer: ```suggestion
``"grandparent"`` is an ancestor of ``"grandparent.parent"``, which is an ancestor of
```

``"grandparent.parent.child"``. A logger inherits the properties of its ancestor logger and

> can define its own.

Reviewer: ```suggestion
can also define its own properties.
```

The {+driver-short+} defines the following logger names to organize different logging events:

.. list-table::

>    :widths: 25 75

Reviewer: maybe make the first column wider so the names don't wrap but the descriptive text does

   :header-rows: 1

   * - Logger Name
     - Description

   * - ``org.mongodb.driver.authenticator``
     - Logs authentication events

   * - ``org.mongodb.driver.client``
     - Logs events related to MongoClient instances

   * - ``org.mongodb.driver.cluster``
     - Logs events related to monitoring of MongoDB deployments

   * - ``org.mongodb.driver.connection``
     - Logs connections and connection pools

   * - ``org.mongodb.driver.connection.tls``
     - Logs TLS events

   * - ``org.mongodb.driver.operation``
     - Logs operations, including logging related to automatic retries

   * - ``org.mongodb.driver.protocol``
     - Logs commands sent to and replies received from MongoDB deployments

   * - ``org.mongodb.driver.uri``
     - Logs connection string parsing events

   * - ``org.mongodb.driver.management``
     - Logs Java Management Extensions (JMX) events

Logger Name Example
~~~~~~~~~~~~~~~~~~~

The following example disables the root logger and sets the log level for the

> ``org.mongodb.driver.connection`` logger to ``DEBUG``. This will cause the driver to log

Reviewer: ```suggestion
``org.mongodb.driver.connection`` logger to ``DEBUG``. This will cause the driver to only log
```

Author: [Style guide entry on "only"](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#o:~:text=Don%27t%20hyphenate.-,only,-Place%20only%20preceding) – "Place _only_ preceding the word or group that it modifies". I _think_ I adhered to that in my version ("messages" being the word getting modified).

Reviewer: maybe put only after messages? it seems a little clunky but i get how style guide rules can do that sometimes. 
Ex. This will cause the driver to log messages only related to.... or log messages related only to
But I think what you have is fine, it just displeases me for some reason but its probably personal preference

Author: I can shift it around!

> only messages related to connections and connection pools at the ``DEBUG`` level or higher.

Reviewer: ```suggestion
messages related to connections and connection pools at the ``DEBUG`` level or higher.
```

      - ``TRACE``

      Set your ``logback.xml`` file to the following:

      .. code-block:: xml

         <configuration>
            <appender name="CONSOLE"
                      class="ch.qos.logback.core.ConsoleAppender">
               <encoder>
                  <pattern>
                        %-4relative [%thread] %-5level %logger{30} - %msg%n
                  </pattern>
               </encoder>
            </appender>
            <root level="INFO">
               <appender-ref ref="CONSOLE" />
            </root>
         </configuration>

      To test that your logger configuration was successful, run the following
      code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none
         :copyable: false

         ...
         1317 [cluster-ClusterId{value='<your cluster id>', description='null'}-<your connection uri>] INFO  org.mongodb.driver.cluster - Discovered replica set primary <your connection uri>
         1568 [main] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:7, serverValue:<server value>}] to <your connection uri>

> 

Reviewer 2: These log messages are no longer emitted by drivers starting with 4.10 as well.

The actual output I observed when running the Kotlin driver is as follows:
```
13:21:29.111 [main] DEBUG org.mongodb.driver.cluster - Starting monitoring for topology with ID ClusterId{value='687ea14948c42b0971fc4036', description='null'}
13:21:29.116 [main] DEBUG org.mongodb.driver.cluster - Updating cluster description to  {type=UNKNOWN, servers=[{address=localhost:27017, type=UNKNOWN, state=CONNECTING}]
13:21:29.117 [main] DEBUG org.mongodb.driver.cluster - Description changed for topology with ID ClusterId{value='687ea14948c42b0971fc4036', description='null'}. Previous description: {type=UNKNOWN, servers=[]. New description: {type=UNKNOWN, servers=[{address=localhost:27017, type=UNKNOWN, state=CONNECTING}]
13:21:29.130 [main] DEBUG org.mongodb.driver.connection - Connection pool created for localhost:27017 using options maxIdleTimeMS=0, minPoolSize=0, maxPoolSize=100, maxConnecting=2, waitQueueTimeoutMS=120000
13:21:29.134 [main] DEBUG org.mongodb.driver.cluster - Starting monitoring for server localhost:27017 in topology with ID ClusterId{value='687ea14948c42b0971fc4036', description='null'}
13:21:29.134 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.cluster - Checking status of localhost:27017
13:21:29.142 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.cluster - Heartbeat started for localhost:27017 on connection with driver-generated ID 1 and server-generated ID null in topology with ID ClusterId{value='687ea14948c42b0971fc4036', description='null'}. Awaited: false
13:21:29.164 [main] INFO  org.mongodb.driver.client - MongoClient with metadata {"driver": {"name": "mongo-java-driver|sync|kotlin", "version": "5.6.0-alpha0-20-g222541cd92-dirty"}, "os": {"type": "Darwin", "name": "Mac OS X", "architecture": "aarch64", "version": "15.5"}, "platform": "Java/JetBrains s.r.o./17.0.9+8-b1166.2|kotlin/1.8.10"} created with settings MongoClientSettings{readPreference=primary, writeConcern=WriteConcern{w=null, wTimeout=null ms, journal=null}, retryWrites=true, retryReads=true, readConcern=ReadConcern{level=null}, credential=null, transportSettings=null, commandListeners=[], codecRegistry=ProvidersCodecRegistry{codecProviders=[ValueCodecProvider{}, BsonValueCodecProvider{}, DBRefCodecProvider{}, DBObjectCodecProvider{}, DocumentCodecProvider{}, CollectionCodecProvider{}, IterableCodecProvider{}, MapCodecProvider{}, GeoJsonCodecProvider{}, GridFSFileCodecProvider{}, Jsr310CodecProvider{}, JsonObjectCodecProvider{}, BsonCodecProvider{}, EnumCodecProvider{}, com.mongodb.client.model.mql.ExpressionCodecProvider@63a12c68, com.mongodb.Jep395RecordCodecProvider@28f3b248, com.mongodb.KotlinCodecProvider@1b1426f4]}, loggerSettings=LoggerSettings{maxDocumentLength=1000}, clusterSettings={hosts=[localhost:27017], srvServiceName=mongodb, mode=SINGLE, requiredClusterType=UNKNOWN, requiredReplicaSetName='null', serverSelector='null', clusterListeners='[]', serverSelectionTimeout='30000 ms', localThreshold='15 ms'}, socketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=0, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, heartbeatSocketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=10000, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, connectionPoolSettings=ConnectionPoolSettings{maxSize=100, minSize=0, maxWaitTimeMS=120000, maxConnectionLifeTimeMS=0, maxConnectionIdleTimeMS=0, maintenanceInitialDelayMS=0, maintenanceFrequencyMS=60000, connectionPoolListeners=[], maxConnecting=2}, serverSettings=ServerSettings{heartbeatFrequencyMS=10000, minHeartbeatFrequencyMS=500, serverMonitoringMode=AUTO, serverListeners='[]', serverMonitorListeners='[]'}, sslSettings=SslSettings{enabled=false, invalidHostNameAllowed=false, context=null}, applicationName='null', compressorList=[], uuidRepresentation=UNSPECIFIED, serverApi=null, autoEncryptionSettings=null, dnsClient=null, inetAddressResolver=null, contextProvider=null, timeoutMS=null}
13:21:29.173 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] INFO  org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=localhost:27017, type=SHARD_ROUTER, cryptd=false, state=CONNECTED, ok=true, minWireVersion=0, maxWireVersion=26, maxDocumentSize=16777216, logicalSessionTimeoutMinutes=30, roundTripTimeNanos=19451542, minRoundTripTimeNanos=0}
13:21:29.174 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.connection - Connection pool ready for localhost:27017
13:21:29.189 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.cluster - Updating cluster description to  {type=SHARDED, servers=[{address=localhost:27017, type=SHARD_ROUTER, roundTripTime=19.5 ms, state=CONNECTED}]
13:21:29.189 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.cluster - Description changed for topology with ID ClusterId{value='687ea14948c42b0971fc4036', description='null'}. Previous description: {type=UNKNOWN, servers=[{address=localhost:27017, type=UNKNOWN, state=CONNECTING}]. New description: {type=SHARDED, servers=[{address=localhost:27017, type=SHARD_ROUTER, roundTripTime=19.5 ms, state=CONNECTED}]
13:21:29.189 [cluster-ClusterId{value='687ea14948c42b0971fc4036', description='null'}-localhost:27017] DEBUG org.mongodb.driver.cluster - Checking status of localhost:27017
13:21:29.202 [main] DEBUG org.mongodb.driver.cluster - Server selection started for operation find with ID 5. Selector: ReadPreferenceServerSelector{readPreference=primary}, topology description: {type=SHARDED, servers=[{address=localhost:27017, type=SHARD_ROUTER, roundTripTime=19.5 ms, state=CONNECTED}]
13:21:29.208 [main] DEBUG org.mongodb.driver.cluster - Server selection succeeded for operation find with ID 5. Selected server: localhost:27017. Selector: ReadPreferenceServerSelector{readPreference=primary}, topology description: {type=SHARDED, servers=[{address=localhost:27017, type=SHARD_ROUTER, roundTripTime=19.5 ms, state=CONNECTED}]
13:21:29.209 [main] DEBUG org.mongodb.driver.connection - Checkout started for connection to localhost:27017
13:21:29.213 [main] DEBUG org.mongodb.driver.connection - Connection created: address=localhost:27017, driver-generated ID=3
13:21:29.218 [main] DEBUG org.mongodb.driver.connection - Connection ready: address=localhost:27017, driver-generated ID=3, established in=5 ms
13:21:29.219 [main] DEBUG org.mongodb.driver.connection - Connection checked out: address=localhost:27017, driver-generated ID=3, duration=10 ms
13:21:29.237 [main] DEBUG org.mongodb.driver.protocol.command - Command "find" started on database "test" using a connection with driver-generated ID 3 and server-generated ID 516757 to localhost:27017. The request ID is 5 and the operation ID is 5. Command: {"find": "test", "filter": {}, "limit": 1, "singleBatch": true, "$db": "test", "lsid": {"id": {"$binary": {"base64": "Bw3a3QK7QAGob8VghbJguw==", "subType": "04"}}}}
13:21:29.244 [main] DEBUG org.mongodb.driver.protocol.command - Command "find" succeeded on database "test" in 8.823416 ms using a connection with driver-generated ID 3 and server-generated ID 516757 to localhost:27017. The request ID is 5 and the operation ID is 5. Command reply: {"cursor": {"firstBatch": [{"_id": {"$oid": "6876d15f70164c1e5a47be02"}, "advance": "$clusterTime"}], "id": 0, "ns": "test.test"}, "ok": 1.0, "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1753129287, "i": 1}}, "signature": {"hash": {"$binary": {"base64": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "subType": "00"}}, "keyId": 0}}, "operationTime": {"$timestamp": {"t": 1753129283, "i": 1}}}
13:21:29.249 [main] DEBUG org.mongodb.driver.operation - Received batch of 1 documents with cursorId 0 from server localhost:27017
13:21:29.251 [main] DEBUG org.mongodb.driver.connection - Connection checked in: address=localhost:27017, driver-generated ID=3
```

I suggest we update the expected output to match the current driver logs. I’ve included the full output from the executed example, but we could trim it to only the relevant parts for clarity.

      - ``FATAL``
      - ``ERROR``
      - ``WARN``
      - ``INFO``
      - ``DEBUG``
      - ``TRACE``
      - ``ALL``

      Set your ``log4j2.xml`` file to the following:

      .. code-block:: xml

         <?xml version="1.0" encoding="UTF-8"?>
         <Configuration status="INFO">
            <Appenders>
               <Console name="Console" target="SYSTEM_OUT">
                     <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
               </Console>
            </Appenders>
            <Loggers>
               <Root level="INFO">
                     <AppenderRef ref="Console"/>
               </Root>
            </Loggers>
         </Configuration>

      To test that your logger configuration was successful, run the following
      code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none
         :copyable: false

         ...
         10:14:57.633 [cluster-ClusterId{value=<your cluster id>, description='null'}-<your connection uri>] INFO  org.mongodb.driver.cluster - Discovered replica set primary <your connection uri>

>          10:14:57.790 [main] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:7, serverValue:<your server value>}] to <your connection uri>

Reviewer 2: These log messages are no longer emitted by drivers starting from 
4.10. 

The actual output I observed when running the Kotlin driver is as follows:

```
221  [main] INFO  org.mongodb.driver.client - MongoClient with metadata {"driver": {"name": "mongo-java-driver|sync", "version": "5.6.0-alpha0-20-g222541cd92-dirty"}, "os": {"type": "Darwin", "name": "Mac OS X", "architecture": "aarch64", "version": "15.5"}, "platform": "Java/JetBrains s.r.o./17.0.9+8-b1166.2"} created with settings MongoClientSettings{readPreference=primary, writeConcern=WriteConcern{w=null, wTimeout=null ms, journal=null}, retryWrites=true, retryReads=true, readConcern=ReadConcern{level=null}, credential=null, transportSettings=null, commandListeners=[], codecRegistry=ProvidersCodecRegistry{codecProviders=[ValueCodecProvider{}, BsonValueCodecProvider{}, DBRefCodecProvider{}, DBObjectCodecProvider{}, DocumentCodecProvider{}, CollectionCodecProvider{}, IterableCodecProvider{}, MapCodecProvider{}, GeoJsonCodecProvider{}, GridFSFileCodecProvider{}, Jsr310CodecProvider{}, JsonObjectCodecProvider{}, BsonCodecProvider{}, EnumCodecProvider{}, com.mongodb.client.model.mql.ExpressionCodecProvider@35399441, com.mongodb.Jep395RecordCodecProvider@4b7dc788, com.mongodb.KotlinCodecProvider@6304101a]}, loggerSettings=LoggerSettings{maxDocumentLength=1000}, clusterSettings={hosts=[localhost:27017], srvServiceName=mongodb, mode=SINGLE, requiredClusterType=UNKNOWN, requiredReplicaSetName='null', serverSelector='null', clusterListeners='[]', serverSelectionTimeout='30000 ms', localThreshold='15 ms'}, socketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=0, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, heartbeatSocketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=10000, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, connectionPoolSettings=ConnectionPoolSettings{maxSize=100, minSize=0, maxWaitTimeMS=120000, maxConnectionLifeTimeMS=0, maxConnectionIdleTimeMS=0, maintenanceInitialDelayMS=0, maintenanceFrequencyMS=60000, connectionPoolListeners=[], maxConnecting=2}, serverSettings=ServerSettings{heartbeatFrequencyMS=10000, minHeartbeatFrequencyMS=500, serverMonitoringMode=AUTO, serverListeners='[]', serverMonitorListeners='[]'}, sslSettings=SslSettings{enabled=false, invalidHostNameAllowed=false, context=null}, applicationName='null', compressorList=[], uuidRepresentation=UNSPECIFIED, serverApi=null, autoEncryptionSettings=null, dnsClient=null, inetAddressResolver=null, contextProvider=null, timeoutMS=null}
241  [main] INFO  org.mongodb.driver.cluster - Waiting for server to become available for operation find with ID 3. Remaining time: 29996 ms. Selector: ReadPreferenceServerSelector{readPreference=primary}, topology description: {type=UNKNOWN, servers=[{address=localhost:27017, type=UNKNOWN, state=CONNECTING}].
242  [cluster-ClusterId{value='687ea2a581c0d42ae6d42552', description='null'}-localhost:27017] INFO  org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=localhost:27017, type=SHARD_ROUTER, cryptd=false, state=CONNECTED, ok=true, minWireVersion=0, maxWireVersion=26, maxDocumentSize=16777216, logicalSessionTimeoutMinutes=30, roundTripTimeNanos=15033750, minRoundTripTimeNanos=0}
```

The behavior is the same with log4j, though the layout differs. Should we update the expected message to reflect the current driver output?

Logger Setup Example
~~~~~~~~~~~~~~~~~~~~

The following example shows how to set up a logger. Select the :guilabel:`Logback` or
:guilabel:`Log4j2` tab to see the corresponding syntax:

.. tabs::

   .. tab:: Logback
      :tabid: Logback-binding

      Select the :guilabel:`Maven` or :guilabel:`Gradle` tab to learn about how to set up
      Logback with your project's dependency management tool:

      .. tabs::

         .. tab:: Maven
            :tabid: maven

            Add the following dependency to your ``pom.xml`` file.

            .. include:: /includes/logging-and-monitoring/logback-maven.rst

         .. tab:: Gradle
            :tabid: gradle

            Add the following dependency to your ``build.gradle`` file.

            .. include:: /includes/logging-and-monitoring/logback-gradle.rst

      Once you have included the preceding dependency, connect to your
      MongoDB instance and retrieve a document with the following code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      The preceding code will ouput a message that resembles the following:

      .. code-block:: none
         :copyable: false

         ...

>          12:14:55.853 [main] DEBUG org.mongodb.driver.connection - Opened connection [connectionId{localValue:3, serverValue:3}] to <MongoDB hostname>

Reviewer 2: This log line is no longer present in newer drivers. We can substitute this line with:

```
12:14:55.833 [main] DEBUG org.mongodb.driver.connection - Checkout started for connection to <MongoDB hostname>
12:14:55.834 [main] DEBUG org.mongodb.driver.connection - Connection created: address=<MongoDB hostname>, driver-generated ID=3
12:14:55.836 [main] DEBUG org.mongodb.driver.connection - Connection ready: address=<MongoDB hostname>, driver-generated ID=3, established in=4 ms
12:14:55.843 [main] DEBUG org.mongodb.driver.connection - Connection checked out: address=<MongoDB hostname>, driver-generated ID=3, duration=9 ms
```

      :tabid: Logback-binding

      Specify Logback configurations in a file named ``logback.xml``.
      You must be able to access this file from your classpath.

      The Logback framework defines the following log levels ordered from most urgent to
      least urgent:

      - ``ERROR``
      - ``WARN``
      - ``INFO``
      - ``DEBUG``
      - ``TRACE``

      Set your ``logback.xml`` file to the following:

      .. code-block:: xml

         <configuration>
            <appender name="CONSOLE"
                      class="ch.qos.logback.core.ConsoleAppender">
               <encoder>
                  <pattern>
                        %-4relative [%thread] %-5level %logger{30} - %msg%n
                  </pattern>
               </encoder>
            </appender>
            <root level="INFO">
               <appender-ref ref="CONSOLE" />
            </root>
         </configuration>

      To test that your logger configuration was successful, run the following
      code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none
         :copyable: false

         ...

>          13:21:29.164 [main] INFO  org.mongodb.driver.client - MongoClient with metadata {"driver": {"name": "mongo-java-driver|sync|kotlin", "version": "5.6.0-alpha0-20-g222541cd92-dirty"}, "os": {"type": "Darwin", "name": "Mac OS X", "architecture": "aarch64", "version": "15.5"}, "platform": "Java/JetBrains s.r.o./17.0.9+8-b1166.2|kotlin/1.8.10"} created with settings MongoClientSettings{readPreference=primary, writeConcern=WriteConcern{w=null, wTimeout=null ms, journal=null}, retryWrites=true, retryReads=true, readConcern=ReadConcern{level=null}, credential=null, transportSettings=null, commandListeners=[], codecRegistry=ProvidersCodecRegistry{codecProviders=[ValueCodecProvider{}, BsonValueCodecProvider{}, DBRefCodecProvider{}, DBObjectCodecProvider{}, DocumentCodecProvider{}, CollectionCodecProvider{}, IterableCodecProvider{}, MapCodecProvider{}, GeoJsonCodecProvider{}, GridFSFileCodecProvider{}, Jsr310CodecProvider{}, JsonObjectCodecProvider{}, BsonCodecProvider{}, EnumCodecProvider{}, com.mongodb.client.model.mql.ExpressionCodecProvider@63a12c68, com.mongodb.Jep395RecordCodecProvider@28f3b248, com.mongodb.KotlinCodecProvider@1b1426f4]}, loggerSettings=LoggerSettings{maxDocumentLength=1000}, clusterSettings={hosts=[localhost:27017], srvServiceName=mongodb, mode=SINGLE, requiredClusterType=UNKNOWN, requiredReplicaSetName='null', serverSelector='null', clusterListeners='[]', serverSelectionTimeout='30000 ms', localThreshold='15 ms'}, socketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=0, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, heartbeatSocketSettings=SocketSettings{connectTimeoutMS=10000, readTimeoutMS=10000, receiveBufferSize=0, proxySettings=ProxySettings{host=null, port=null, username=null, password=null}}, connectionPoolSettings=ConnectionPoolSettings{maxSize=100, minSize=0, maxWaitTimeMS=120000, maxConnectionLifeTimeMS=0, maxConnectionIdleTimeMS=0, maintenanceInitialDelayMS=0, maintenanceFrequencyMS=60000, connectionPoolListeners=[], maxConnecting=2}, serverSettings=ServerSettings{heartbeatFrequencyMS=10000, minHeartbeatFrequencyMS=500, serverMonitoringMode=AUTO, serverListeners='[]', serverMonitorListeners='[]'}, sslSettings=SslSettings{enabled=false, invalidHostNameAllowed=false, context=null}, applicationName='null', compressorList=[], uuidRepresentation=UNSPECIFIED, serverApi=null, autoEncryptionSettings=null, dnsClient=null, inetAddressResolver=null, contextProvider=null, timeoutMS=null}

Reviewer 2: The logs provided here are mixed with DEBUG level, however, this section is about INFO level.

Currently, the driver produces the following logs:
```
317  [main] INFO  org.mongodb.driver.client - MongoClient with metadata {"driver": {"name": "<driver-name>", "version": "<driver-version>"}, "os": {"type": "<os-type>", "name": "<os-name>", "architecture": "<architecture>", "version": "<os-version>"}, "platform": "<platform-info>"} created with settings <MongoClientSettings>
345  [cluster-ClusterId{value='<cluster-id>', description='null'}-<MongoDB hostname>] INFO  org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=<MongoDB hostname>, type=<server-type>, cryptd=false, state=CONNECTED, ok=true, minWireVersion=0, maxWireVersion=<max-wire-version>, maxDocumentSize=16777216, logicalSessionTimeoutMinutes=30, roundTripTimeNanos=<round-trip-nanos>, minRoundTripTimeNanos=<min-round-trip-nanos}
```
I substituted settings with placeholders.

The same applies for log4j example:

```
21:19:25.696 [main] INFO  org.mongodb.driver.client - MongoClient with metadata {"driver": {"name": "<driver-name>", "version": "<driver-version>"}, "os": {"type": "<os-type>", "name": "<os-name>", "architecture": "<architecture>", "version": "<os-version>"}, "platform": "<platform-info>"} created with settings <MongoClientSettings>
21:19:25.710 [cluster-ClusterId{value='<cluster-id>', description='null'}-<MongoDB hostname>] INFO  org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=<MongoDB hostname>, type=<server-type>, cryptd=false, state=CONNECTED, ok=true, minWireVersion=0, maxWireVersion=<max-wire-version>, maxDocumentSize=16777216, logicalSessionTimeoutMinutes=30, roundTripTimeNanos=<round-trip-nanos>, minRoundTripTimeNanos=<min-round-trip-nanos}
```

      Specify Log4j2 configurations in a file named ``log4j2.xml``.
      You must be able to access this file from your classpath.

      The Log4j2 framework defines the following log levels, ordered from most urgent to
      least urgent:

      - ``FATAL``
      - ``ERROR``
      - ``WARN``
      - ``INFO``
      - ``DEBUG``
      - ``TRACE``
      - ``ALL``

      Set your ``log4j2.xml`` file to the following:

      .. code-block:: xml

         <?xml version="1.0" encoding="UTF-8"?>
         <Configuration status="INFO">
            <Appenders>
               <Console name="Console" target="SYSTEM_OUT">
                     <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
               </Console>
            </Appenders>
            <Loggers>
               <Root level="INFO">
                     <AppenderRef ref="Console"/>
               </Root>
            </Loggers>
         </Configuration>

      To test that your logger configuration was successful, run the following
      code:

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none

>          :copyable: false

Reviewer 2: Could you clarify why the log example was removed? We could use log4j example from this comment: https://github.com/mongodb/docs-kotlin-sync/pull/115#discussion_r2224157733

Author: This was a pure miss on my end, will fix!

   * - ``org.mongodb.driver.protocol``
     - Logs commands sent to and replies received from MongoDB deployments

   * - ``org.mongodb.driver.uri``
     - Logs connection string parsing events

   * - ``org.mongodb.driver.management``
     - Logs Java Management Extensions (JMX) events

Logger Name Example
~~~~~~~~~~~~~~~~~~~

The following example disables the root logger and sets the log level for the
``org.mongodb.driver.connection`` logger to ``DEBUG``. This will cause the driver to log
messages only related to connections and connection pools at the ``DEBUG`` level or higher.
Select the :guilabel:`Logback` or :guilabel:`Log4j2` tab to see the corresponding syntax:

.. tabs::

   .. tab:: Logback
      :tabid: Logback-binding

      Set your ``logback.xml`` file to the following:

      .. code-block:: xml
         :emphasize-lines: 10

         <configuration>
            <appender name="CONSOLE"
                      class="ch.qos.logback.core.ConsoleAppender">
               <encoder>
                  <pattern>
                        %-4relative [%thread] %-5level %logger{30} - %msg%n
                  </pattern>
               </encoder>
            </appender>

>             <logger name="org.mongodb.driver.connection" level="INFO" additivity="true"/>

Reviewer 2: It is mentioned above ```
"the following example disables the root logger and sets the log level for the
``org.mongodb.driver.connection`` logger to ``DEBUG``"
```, thus we have to set `DEBUG` level in configuration, not INFO.

```suggestion
            <logger name="org.mongodb.driver.connection" level="DEBUG" additivity="true"/>
```

            <root level="OFF">
               <appender-ref ref="CONSOLE" />
            </root>
         </configuration>

      To test that your logger configuration was successful, run the following
      code.

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following.

      .. code-block:: none
         :copyable: false

         ...
         829  [cluster-rtt-ClusterId{value='<some value>', description='null'}-<your connection URI>] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:2, serverValue:<your server value>}] to <your connection uri>

>          977  [main] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:7, serverValue:<your server value>}] to <your connection uri>

Reviewer 2: These logs are no longer present and don’t align with the text above, which states that DEBUG level is enabled.

We should update the example to reflect the configuration suggested in the comment above: [discussion_r2224312327](https://github.com/mongodb/docs-kotlin-sync/pull/115#discussion_r2224312327)

When that configuration is applied, the driver produces the following logs - I suggest we use those in the example. What do you think?

 
```
255  [main] DEBUG org.mongodb.driver.connection - Connection pool created for <MongoDB hostname> using options <connection-pool-options>  
301  [cluster-ClusterId{value='<cluster-id>', description='null'}-<MongoDB hostname>] DEBUG org.mongodb.driver.connection - Connection pool ready for <MongoDB hostname>  
321  [main] DEBUG org.mongodb.driver.connection - Checkout started for connection to <MongoDB hostname>  
323  [main] DEBUG org.mongodb.driver.connection - Connection created: address=<MongoDB hostname>, driver-generated ID=3  
335  [main] DEBUG org.mongodb.driver.connection - Connection ready: address=<MongoDB hostname>, driver-generated ID=3, established in=<connect-duration> ms  
336  [main] DEBUG org.mongodb.driver.connection - Connection checked out: address=<MongoDB hostname>, driver-generated ID=3, duration=<checkout-duration> ms  
363  [main] DEBUG org.mongodb.driver.connection - Connection checked in: address=<MongoDB hostname>, driver-generated ID=3 
```

   .. tab:: Log4j2
      :tabid: Log4j2-binding

      Set your ``log4j2.xml`` file to the following:

      .. code-block:: xml
         :emphasize-lines: 9

         <?xml version="1.0" encoding="UTF-8"?>
         <Configuration status="INFO">
            <Appenders>
               <Console name="Console" target="SYSTEM_OUT">
                     <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
               </Console>
            </Appenders>
            <Loggers>

>                <Logger name="org.mongodb.driver.connection" level="INFO"/>

Reviewer 2: It is mentioned above "the following example disables the root logger and sets the log level for the ``org.mongodb.driver.connection`` logger to ``DEBUG``", thus we have to set DEBUG level in configuration, not INFO.

```suggestion
               <Logger name="org.mongodb.driver.connection" level="DEBUG"/>
```

               <Root level="OFF">
                     <AppenderRef ref="Console"/>
               </Root>
            </Loggers>
         </Configuration>

      To test that your logger configuration was successful, run the following
      code.

      .. include:: /includes/logging-and-monitoring/logging.rst

      This code produces output that resembles the following:

      .. code-block:: none
         :copyable: false

         ...
         15:40:23.005 [cluster-ClusterId{value='<some value>', description='null'}-<your connection uri>] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:3, serverValue:<your server value>}] to <your connection uri>

>          15:40:23.159 [main] INFO  org.mongodb.driver.connection - Opened connection [connectionId{localValue:7, serverValue:<your server value>}] to <your connection uri>

Reviewer 2: These logs are no longer present and don’t align with the text above, which states that DEBUG level is enabled.

We should update the example to reflect the configuration suggested in the comment above:  [discussion_r2224256906](https://github.com/mongodb/docs-kotlin-sync/pull/115#discussion_r2224256906)

When that configuration is applied, the driver produces the following logs - I suggest we use those in the example. What do you think?

```
21:27:17.035  [main] DEBUG org.mongodb.driver.connection - Connection pool created for <MongoDB hostname> using options <connection-pool-options>  
21:27:17.058  [cluster-ClusterId{value='<cluster-id>', description='null'}-<MongoDB hostname>] DEBUG org.mongodb.driver.connection - Connection pool ready for <MongoDB hostname>  
21:27:17.064  [main] DEBUG org.mongodb.driver.connection - Checkout started for connection to <MongoDB hostname>  
21:27:17.069  [main] DEBUG org.mongodb.driver.connection - Connection created: address=<MongoDB hostname>, driver-generated ID=3  
21:27:17.075  [main] DEBUG org.mongodb.driver.connection - Connection ready: address=<MongoDB hostname>, driver-generated ID=3, established in=<connect-duration> ms  
21:27:17.075  [main] DEBUG org.mongodb.driver.connection - Connection checked out: address=<MongoDB hostname>, driver-generated ID=3, duration=<checkout-duration> ms  
21:27:17.086  [main] DEBUG org.mongodb.driver.connection - Connection checked in: address=<MongoDB hostname>, driver-generated ID=3 
```
